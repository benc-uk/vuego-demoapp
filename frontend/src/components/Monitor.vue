<!--
Monitoring page, fetches metrics data from API
----------------------------------------------
Ben C, April 2018, Updated for Vue3 2021
-->

<template>
  <div class="card border-light bg-dark mb-3 rounded-3">
    <h4 class="card-header bg-info">
      <i class="fas fa-tachometer-alt fa-fw me-3"></i>
      Monitoring
    </h4>

    <div class="card-body">
      <div v-if="error" class="alert text-dark alert-warning">
        <h4>There was a problem ðŸ˜¥</h4>
        <div class="errmsg">
          {{ error }}
        </div>
      </div>

      <spinner v-if="!metrics && !error" />

      <div v-if="metrics" class="container" fluid>
        <div class="row">
          <div class="col border-light">
            <dial :value="cpu" :title="'CPU Load'" :percentage="true" />
          </div>
          <div class="col">
            <dial :value="mem" :title="'Memory Used'" :percentage="true" />
          </div>
        </div>
        <div class="row">
          <div class="col">
            <dial :value="disk" :title="'Disk Used'" :percentage="true" />
          </div>
          <div class="col">
            <dial :value="net" :title="'Net I/O'" :percentage="false" />
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import apiMixin from '../mixins/apiMixin.js'
import Spinner from './Spinner.vue'
import Dial from './Dial.vue'

let prevNetBytes
let refreshId

export default {
  components: {
    Spinner,
    Dial
  },

  // Adds functions to call the API
  mixins: [apiMixin],

  data: function () {
    return {
      metrics: null,
      prevNetBytes: null,
      error: null
    }
  },

  computed: {
    cpu: function () {
      if (!this.metrics) {
        return 0
      }
      return this.metrics.cpuPerc
    },

    mem: function () {
      if (!this.metrics) {
        return 0
      }
      return (this.metrics.memUsed / this.metrics.memTotal) * 100
    },

    disk: function () {
      if (!this.metrics) {
        return 0
      }
      return 100 - (this.metrics.diskFree / this.metrics.diskTotal) * 100
    },

    net: function () {
      const newTot = this.metrics.netBytesSent + this.metrics.netBytesRecv
      const delta = newTot - prevNetBytes
      if (prevNetBytes) {
        prevNetBytes = newTot
        // Scaling factor here is mostly arbitrary to get nice looking numbers
        return delta / 1000
      } else {
        prevNetBytes = newTot
        return 0
      }
    }
  },

  created() {
    this.update()
    refreshId = setInterval(this.update, 2500)
  },

  beforeUnmount() {
    clearInterval(refreshId)
  },

  methods: {
    // Update the data with an API call
    update: async function () {
      this.error = null
      try {
        this.metrics = await this.apiGetMetrics()
      } catch (err) {
        this.error = err
      }
    }
  }
}
</script>
