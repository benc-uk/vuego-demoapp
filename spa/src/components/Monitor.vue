<!--
Monitoring page, fetches metrics data from API
----------------------------------------------
Ben C, April 2018
-->

<template>
  <div class="card">
    <div class="card-header bg-success h3"><fa icon="tachometer-alt" />&nbsp; Monitoring</div>
    <div class="card-body">
      <b-alert v-if="error" show variant="warning">
        <h4>There was a problem ðŸ˜¥</h4>
        <div class="errmsg">
          {{ error }}
        </div>
      </b-alert>

      <spinner v-if="!metrics && !error" />

      <b-container v-if="metrics" fluid>
        <b-row align-h="around">
          <b-col md>
            <dial :value="cpu" :title="'CPU Load'" :percentage="true" />
          </b-col>
          <b-col md>
            <dial :value="mem" :title="'Memory Used'" :percentage="true" />
          </b-col>
        </b-row>
        <b-row align-h="around">
          <b-col md>
            <dial :value="disk" :title="'Disk Used'" :percentage="true" />
          </b-col>
          <b-col md>
            <dial :value="net" :title="'Net I/O'" :percentage="false" />
          </b-col>
        </b-row>
      </b-container>
    </div>
  </div>
</template>

<script>
import apiMixin from '../mixins/apiMixin.js'
import Spinner from './Spinner.vue'
import Dial from './Dial.vue'

let prevNetBytes
let refreshId

export default {
  components: {
    Spinner,
    Dial
  },

  // Adds functions to call the API
  mixins: [apiMixin],

  data: function() {
    return {
      metrics: null,
      prevNetBytes: null,
      error: null
    }
  },

  computed: {
    cpu: function() {
      if (!this.metrics) {
        return 0
      }
      return this.metrics.cpuPerc
    },

    mem: function() {
      if (!this.metrics) {
        return 0
      }
      return (this.metrics.memUsed / this.metrics.memTotal) * 100
    },

    disk: function() {
      if (!this.metrics) {
        return 0
      }
      return 100 - (this.metrics.diskFree / this.metrics.diskTotal) * 100
    },

    net: function() {
      let newTot = this.metrics.netBytesSent + this.metrics.netBytesRecv
      let delta = newTot - prevNetBytes
      if (prevNetBytes) {
        prevNetBytes = newTot
        // Scaling factor here is mostly arbitrary to get nice looking numbers
        return delta / 1000
      } else {
        prevNetBytes = newTot
        return 0
      }
    }
  },

  created() {
    this.update()
    refreshId = setInterval(this.update, 2500)
  },

  beforeDestroy() {
    clearInterval(refreshId)
  },

  methods: {
    // Update the data with an API call
    update: async function() {
      this.error = null
      try {
        this.metrics = await this.apiGetMetrics()
      } catch (err) {
        this.error = err
      }
    }
  }
}
</script>

<style scoped>
.card-body {
  background-color: #ddd;
}
</style>
