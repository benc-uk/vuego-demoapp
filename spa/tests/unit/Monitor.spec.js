import { createLocalVue, shallowMount } from '@vue/test-utils'
import Monitor from '@/components/Monitor.vue'
import flushPromises from 'flush-promises'

const localVue = createLocalVue()

const mockMixin = {
  methods: {
    apiGetMetrics() {
      return new Promise((resolve) => {
        resolve({
          memTotal: 300,
          memUsed: 150,
          cpuPerc: 42,
          diskTotal: 500,
          diskFree: 200,
          netBytesSent: 300,
          netBytesRecv: 300
        })
      })
    }
  }
}

describe('Monitor.vue', () => {
  it('renders monitor dials', async () => {
    const wrapper = shallowMount(Monitor, { localVue, mixins: [mockMixin] })
    await flushPromises()
    expect(wrapper).toMatchSnapshot()
  })
})
