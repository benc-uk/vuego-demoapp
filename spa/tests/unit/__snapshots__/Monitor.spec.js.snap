// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Monitor.vue renders monitor dials 1`] = `
<div class="card">
  <div class="card-header bg-success h3">
    <fa icon="tachometer-alt"></fa>&nbsp; Monitoring
  </div>
  <div class="card-body">
    <!---->
    <!---->
    <b-container fluid="">
      <b-row align-h="around">
        <b-col md="">
          <dial-stub value="42" title="CPU Load" percentage="true"></dial-stub>
        </b-col>
        <b-col md="">
          <dial-stub value="50" title="Memory Used" percentage="true"></dial-stub>
        </b-col>
      </b-row>
      <b-row align-h="around">
        <b-col md="">
          <dial-stub value="60" title="Disk Used" percentage="true"></dial-stub>
        </b-col>
        <b-col md="">
          <dial-stub value="0" title="Net I/O"></dial-stub>
        </b-col>
      </b-row>
    </b-container>
  </div>
</div>
`;
